# 反爬：瀏覽器標頭與基本資訊


* 了解「檢查 HTTP 標頭檔」的反爬蟲機制
* 「檢查 HTTP 標頭檔」反爬蟲的因應策略


## 作業目標

* 尋找一個網站是有作 Header 檢查的（Hint: 有加跟沒有加回傳結果不一樣）
* 用上述網站說明該如何判斷 Header 中必須加上哪些資料才會正確


### 尋找一個網站是有作 Header 檢查的（Hint: 有加跟沒有加回傳結果不一樣）

In [1]:
import requests
r = requests.get('https://makezine.com/')
response = r.text

response

'<html>\n<head><title>403 Forbidden</title></head>\n<body>\n<center><h1>403 Forbidden</h1></center>\n<hr><center>nginx</center>\n</body>\n</html>\n'

### 用上述網站說明該如何判斷 Header 中必須加上哪些資料才會正確


In [2]:
import requests
headers = {'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.89 Safari/537.36'}
r = requests.get('https://makezine.com/', headers=headers)
response = r.text

response[:2000]

'<!DOCTYPE html>\n<html lang="en" xmlns:fb="http://ogp.me/ns/fb#" xmlns:fb="https://www.facebook.com/2008/fbml">\n<head>\n<meta charset="utf-8">\n<meta http-equiv="X-UA-Compatible" content="IE=edge">\n<meta name="viewport" content="width=device-width, initial-scale=1">\n<title>Make: DIY Projects and Ideas for Makers | </title>\n<meta name="twitter:widgets:csp" content="on">\n<meta name="p:domain_verify" content="c4e1096cb904ca6df87a2bb867715669">\n<meta name="google-site-verification" content="eAyS6vj89QDOsofiDC4sB-2YR3YNg_VGCYiQnWm2mqc" />\n<meta name="norton-safeweb-site-verification" content="4g4w71jm7qt9e7ghe2dxdhiq0mnnkwom6ue80rdet53q3figx8ooxrffbgkkl9kzo3qi85l2j-txt-fh8w-p2z5769ht01z8s6sxq3-8r7cojmimgp00homsjjjv96ww" />\n<meta property="fb:admins" content="1612614584" />\n<link rel="apple-touch-icon" sizes="180x180" href="https://1abxf1rh6g01lhm2riyrt55k-wpengine.netdna-ssl.com/apple-touch-icon.png">\n<link rel="icon" type="image/png" href="https://1abxf1rh6g01lhm2riyrt55k-wpengi

沒有加入**headers**時回傳403，經搜尋 [HTTP 403](https://zh.wikipedia.org/wiki/HTTP_403) 代表「沒有權限存取此站，伺服器收到請求但拒絕提供服務」。

打開"開發人員工具"F12 > Network，可以查看 Request 帶的 Headers。
![開發人員工具](./headers.png)

並嘗試從最有可能的 User-Agent 參數開始帶入，如果不行，則再嘗試其他參數。